### PR TITLE
Parallel seeds

### PR DESCRIPTION
**Pull request by topepo**
We now generate L'Ecuyer-CMRG seeds that are matched to resamples. These are set for sequential and parallel execution.

### CODE REVIEW

File: R/parallel_new.R [hunk 1]
```diff
@@ -312,8 +338,7 @@ loop_call <-
       future_opts <- list(
         future.label = "tune-grid-%d",
         future.stdout = TRUE,
-        future.seed = TRUE, # TODO <- this line may change
-        future.packages = quote(load_pkgs)
+        future.seed = NULL
```
Comment (topepo) [inline R/parallel_new.R:340]: We are setting our own so don't yell at me bro 😄

File: R/parallel_new.R [hunk 2]
```diff
@@ -152,6 +152,32 @@ choose_framework <- function(
...
+  on.exit({
+    RNGkind(orig_kind)
+    assign(".Random.seed", orig_state, globalenv())
+  })
+
+  # Set to type used for multiple streams
+  RNGkind("L'Ecuyer-CMRG")
+
+  # Capture the seed to make more seeds.
+  runif(1)
```
Comment (topepo) [inline R/parallel_new.R]: Do we need to spend a random number? 🤔

File: R/loop_over_all_stages.R [hunk 3]
```diff
@@ -295,9 +299,15 @@ loop_over_all_stages <- function(resamples, grid, static) {
   return_tbl <- vctrs::vec_cbind(return_tbl, split_labs)
 
   if (static$control$save_pred) {
-    return_tbl$.predictions <- list(add_configs(pred_reserve, static, config_tbl))
+    return_tbl$.predictions <- list(add_configs(
```
Comment (topepo) [inline R/loop_over_all_stages.R:302]: 🤢

File: NEWS.md [hunk 4]
```diff
@@ -6,6 +6,8 @@
 
 * Introduced support for parallel processing with mirai in addition to the currently supported framework future. See `?parallelism` to learn more (#1028).
 
+* Sequential and parallel processing all use the same L'Ecuyer-CMRG seeds (conditional on `parallel_over`) (#1033).
```
Comment (simonpcouch) [inline NEWS.md:9]: As a user, I might wonder what this means for tuning results generated with tune 1.3.0 or prior. The seeds in those versions for a given resample will be different than those for this version, right?

File: R/loop_over_all_stages.R [hunk 5]
```diff
@@ -295,9 +299,15 @@ loop_over_all_stages <- function(resamples, grid, static) {
...
+      pred_reserve,
+      static,
+      config_tbl
+    ))
   }
 
+  # Put the seed back where we found it
+  assign(".Random.seed", orig_seed, envir = globalenv())
```
Comment (simonpcouch) [inline R/loop_over_all_stages.R]:
```suggestion
```

I believe we want this in a `withr::defer()` (or `on.exit()`) in case something goes wrong in the function body

File: R/loop_over_all_stages.R [hunk 6]
```diff
@@ -6,6 +6,11 @@
 loop_over_all_stages <- function(resamples, grid, static) {
   # Initialize some objects
 
+  orig_seed <- .Random.seed
+  # Set seed within the worker process
+  assign(".Random.seed", resamples$.seeds[[1]], envir = .GlobalEnv)
+  resamples$.seeds <- NULL
+
```
Comment (simonpcouch) [inline R/loop_over_all_stages.R]:
```suggestion
  withr::defer(assign(".Random.seed", orig_seed, envir = .GlobalEnv))
```

File: R/parallel_new.R (hunk 2 above)
Comment (simonpcouch) [reply R/parallel_new.R]: From what I can tell, this line can just be removed? If you do think we need to "spend" one, as you say, let's add a comment to clarify the edge case that this covers and link to this PR with `(#1033)`

File: tests/testthat/test-parallel.R [hunk 7]
```diff
@@ -185,3 +185,121 @@ test_that("loop execution code", {
...
+
+  future_mtr <- collect_metrics(future_res)
+
+  expect_equal(seq_mtr, future_mtr)
+
+  future::plan("sequential")
+})
+
+
+test_that("generating parallel seeds does not affect the space-time continuum", {
```
Comment (simonpcouch) [inline tests/testthat/test-parallel.R:281]: 👾🛰️🚀

File: tests/testthat/test-parallel.R [hunk 8]
```diff
@@ -185,3 +185,121 @@ test_that("loop execution code", {
     )
   )
 })
+
+
+test_that("same results using mirai", {
```
Comment (simonpcouch) [inline tests/testthat/test-parallel.R:190]: This is so awesome

File: NEWS.md (hunk 4 above)
Comment (topepo) [reply NEWS.md:9]: Yes. I updated the news in the next PR to note the lack of reproducibility between versions.
